Add initial implementation Workloads SapVirtualInstances SDK

jaskisin opened the pull request:
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

Commit:
Add initial implementation for Azure.ResourceManager.SapVirtualInstances SDK

ajaygit158 commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Azure.ResourceManager.SapVirtualInstances.csproj:
```diff
@@ -0,0 +1,8 @@
+    <Version>1.0.0-beta.1</Version>
+    <PackageId>Azure.ResourceManager.SapVirtualInstances</PackageId>
+    <Description>Azure Resource Manager client SDK for Azure resource provider Workloads.</Description>
+    <PackageTags>azure;management;arm;resource manager;workloads</PackageTags>
```
Lets add "acss", "sap", "sapvirtualinstances" as additional tags

Commit:
Refactor SapVirtualInstanceMetadataTests to use a centralized location variable for AzureLocation

Commit:
Update assets.json to include specific tag for Azure.ResourceManager.SapVirtualInstances

jaskisin commented:
Kindly note that we are planning this merge for January release train

Commit:
Update PackageTags in project file to include additional tags for SAP Virtual Instances

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/CHANGELOG.md:
```diff
@@ -0,0 +1,19 @@
+
+
+
+### Features Added
```
need to update change log entry to contain what's provided in the first release

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/ApplicationServerConfiguration.cs:
```diff
@@ -0,0 +1,88 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Gets or sets the application server configuration. </summary>
+    public partial class ApplicationServerConfiguration
```
The name is too generic. Consider appending RP prefix or renaming it.
- use `prepend-rp-prefix` to automatically append PR name. See example [here](https://github.com/Azure/azure-sdk-for-net/blob/176fd476588e8e61ebce747d66714cb39947cae5/sdk/azurestackhci/Azure.ResourceManager.Hci/src/autorest.md?plain=1#L52-L57).
- use `rename-mapping` to rename a model. See example [here](https://github.com/Azure/azure-sdk-for-net/blob/176fd476588e8e61ebce747d66714cb39947cae5/sdk/azurestackhci/Azure.ResourceManager.Hci/src/autorest.md?plain=1#L58-L76)

archerzz replied:
Some of the model names start with `SapApplicationServer` which looks better than `ApplicationServer`

jaskisin replied:
This is also fixed

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/ApplicationServerConfiguration.cs:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        /// <summary> The subnet id. </summary>
+        public string SubnetId { get; set; }
```
What's the format of this property value?
- If it's a GUID, use type [Guid](https://learn.microsoft.com/en-us/dotnet/api/system.guid?view=net-6.0). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/maps/Azure.ResourceManager.Maps/src/autorest.md?plain=1#L25).
- If it's a resource identifier, use type [ResourceIdentifier](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/ResourceIdentifier.cs). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/search/Azure.Search.Documents/src/autorest.md?plain=1#L106-L115).

jaskisin replied:
Thanks @archerzz , I am using tspconfig for generating the csharp code. the above-mentioned example is for autorest which is not getting accepted along with tspconfig.

Can you please share examples for typespec?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/ApplicationServerVmDetails.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary> Defines the type of application server VM. </summary>
+        public ApplicationServerVirtualMachineType? Type { get; }
+        /// <summary> The virtual machine id. </summary>
+        public string VirtualMachineId { get; }
```
What's the format of this property value?
- If it's a GUID, use type [Guid](https://learn.microsoft.com/en-us/dotnet/api/system.guid?view=net-6.0). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/maps/Azure.ResourceManager.Maps/src/autorest.md?plain=1#L25).
- If it's a resource identifier, use type [ResourceIdentifier](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/ResourceIdentifier.cs). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/search/Azure.Search.Documents/src/autorest.md?plain=1#L106-L115).

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/CentralServerConfiguration.cs:
```diff
@@ -0,0 +1,88 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Gets or sets the central server configuration. </summary>
+    public partial class CentralServerConfiguration
```
Name is too generic

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/CentralServerConfiguration.cs:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        /// <summary> The subnet id. </summary>
+        public string SubnetId { get; set; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DatabaseConfiguration.cs:
```diff
@@ -0,0 +1,106 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Gets or sets the database configuration. </summary>
+    public partial class DatabaseConfiguration
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DatabaseVmDetails.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        /// <summary> The virtual machine id. </summary>
+        public string VirtualMachineId { get; }
```
What's the format of this property value?
- If it's a GUID, use type [Guid](https://learn.microsoft.com/en-us/dotnet/api/system.guid?view=net-6.0). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/maps/Azure.ResourceManager.Maps/src/autorest.md?plain=1#L25).
- If it's a resource identifier, use type [ResourceIdentifier](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/ResourceIdentifier.cs). See example [here](https://github.com/Azure/azure-sdk-for-net/blob/bd953fc7a97bf617c892ce6314b50126b4021a4f/sdk/search/Azure.Search.Documents/src/autorest.md?plain=1#L106-L115).

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DeployerVmPackages.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary> The URL to the deployer VM packages file. </summary>
+        public string Uri { get; set; }
+        /// <summary> The deployer VM packages storage account id. </summary>
+        public string StorageAccountId { get; set; }
```
What's the ID format?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DeploymentConfiguration.cs:
```diff
@@ -0,0 +1,59 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Deployment Configuration. </summary>
+    public partial class DeploymentConfiguration : SapConfiguration
```
name is too generic

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DiscoveryConfiguration.cs:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        /// <summary> The virtual machine ID of the Central Server. </summary>
+        public string CentralServerVmId { get; set; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DiscoveryConfiguration.cs:
```diff
@@ -0,0 +1,43 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Discovery Details. </summary>
+    public partial class DiscoveryConfiguration : SapConfiguration
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DiskDetails.cs:
```diff
@@ -0,0 +1,95 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> The supported disk size details for a disk type. </summary>
+    public partial class DiskDetails
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/DiskSkuName.cs:
```diff
@@ -0,0 +1,66 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Defines the disk sku name. </summary>
+    public readonly partial struct DiskSkuName : IEquatable<DiskSkuName>
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/EnqueueReplicationServerType.cs:
```diff
@@ -0,0 +1,51 @@
+            _value = value ?? throw new ArgumentNullException(nameof(value));
+        }
+
+        private const string EnqueueReplicator1Value = "EnqueueReplicator1";
```
Are those names expected?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/ErrorInformation.cs:
```diff
@@ -0,0 +1,74 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Error definition. </summary>
+    public partial class ErrorInformation
```
Generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/HighAvailabilitySoftwareConfiguration.cs:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        /// <summary> The fencing client id. </summary>
+        public string FencingClientId { get; set; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/ImageReference.cs:
```diff
@@ -0,0 +1,82 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Specifies information about the image to use. You can specify information about platform images, marketplace images, or virtual machine images. This element is required when you want to use a platform image, marketplace image, or virtual machine image, but is not used in other creation operations. NOTE: Image reference publisher and offer can only be set when you create the scal ...
+    public partial class ImageReference
```
Generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/LinuxConfiguration.cs:
```diff
@@ -0,0 +1,54 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Specifies the Linux operating system settings on the virtual machine. For a list of supported Linux distributions, see [Linux on Azure-Endorsed Distributions](https://docs.microsoft.com/azure/virtual-machines/linux/endorsed-distros). </summary>
+    public partial class LinuxConfiguration : OSConfiguration
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/MountFileShareConfiguration.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        /// <summary> The fileshare resource ID. </summary>
+        public string Id { get; set; }
```
What are the formats of those IDs?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/OperationStatusResult.cs:
```diff
@@ -0,0 +1,104 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> The current status of an async operation. </summary>
+    public partial class OperationStatusResult
```
name is generic

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/OperationStatusResult.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        /// <summary> Fully qualified ID for the async operation. </summary>
+        public string Id { get; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/OSConfiguration.cs:
```diff
@@ -0,0 +1,69 @@
+    /// Please note <see cref="OSConfiguration"/> is the base class. According to the scenario, a derived class of the base class might need to be assigned here, or this property needs to be casted to one of the possible derived classes.
+    /// The available derived classes include <see cref="LinuxConfiguration"/> and <see cref="WindowsConfiguration"/>.
+    /// </summary>
+    public abstract partial class OSConfiguration
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/SapInstallWithoutOSConfigSoftwareConfiguration.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary> The URL to the SAP Build of Materials(BOM) file. </summary>
+        public string BomUri { get; set; }
+        /// <summary> The SAP bits storage account id. </summary>
+        public string SapBitsStorageAccountId { get; set; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/SingleServerConfiguration.cs:
```diff
@@ -0,0 +1,102 @@
+        /// <summary> The database type. </summary>
+        public SapDatabaseType? DatabaseType { get; set; }
+        /// <summary> The subnet id. </summary>
+        public string SubnetId { get; set; }
```
What's the format of this ID?

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/SoftwareConfiguration.cs:
```diff
@@ -0,0 +1,69 @@
+    /// Please note <see cref="SoftwareConfiguration"/> is the base class. According to the scenario, a derived class of the base class might need to be assigned here, or this property needs to be casted to one of the possible derived classes.
+    /// The available derived classes include <see cref="ExternalInstallationSoftwareConfiguration"/>, <see cref="SapInstallWithoutOSConfigSoftwareConfiguration"/> and <see cref="ServiceInitiatedSoftwareConfiguration"/>.
+    /// </summary>
+    public abstract partial class SoftwareConfiguration
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/SshKeyPair.cs:
```diff
@@ -0,0 +1,69 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> The SSH Key-pair used to authenticate with the VM. The key needs to be at least 2048-bit and in ssh-rsa format. For creating ssh keys, see [Create SSH keys on Linux and Mac for Linux VMs in Azure](https://docs.microsoft.com/azure/virtual-machines/linux/create-ssh-keys-detailed). </summary>
+    public partial class SshKeyPair
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/SshPublicKey.cs:
```diff
@@ -0,0 +1,65 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Contains information about SSH certificate public key and the path on the Linux VM where the public key is placed. </summary>
+    public partial class SshPublicKey
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/StartContent.cs:
```diff
@@ -0,0 +1,65 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Start SAP instance(s) request body. </summary>
+    public partial class StartContent
```
genric name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/StopContent.cs:
```diff
@@ -0,0 +1,69 @@
+namespace Azure.ResourceManager.SapVirtualInstances.Models
+{
+    /// <summary> Stop SAP instance(s) request body. </summary>
+    public partial class StopContent
```
generic name

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/src/Generated/Models/WindowsConfiguration.cs:
```diff
@@ -0,0 +1,30 @@
+using System;
+using System.Collections.Generic;
+
+namespace Azure.ResourceManager.SapVirtualInstances.Models
```
generic name

Commit:
Release v1.0.0 with SAP configuration updates

Updated CHANGELOG.md for v1.0.0 release, marking the first stable release of the Workloads Management library. This includes separation of the Dotnet SDK Package, adherence to new Azure SDK guidelines, and various core capabilities.

Refactored `Sample_SapApplicationServerInstanceResource.cs`, `Sample_SapCentralServerInstanceResource.cs`, and `Sample_SapDatabaseInstanceResource.cs` to replace `StartContent` with `SapStartContent`.

Updated `Sample_SapVirtualInstanceCollection.cs` to align with new naming conventions and subscription details, including renaming classes and properties to reflect SAP context.

Introduced new classes and serialization logic for handling SAP configurations in Azure Resource Manager, including `SapApplicationServerConfiguration`, `SapCentralServerConfiguration`, `SapDatabaseConfiguration`, `SapDeploymentConfiguration`, `SapDiscoveryConfiguration`, `SapDiskDetails`, `SapImageReference`, `SapLinuxConfiguration`, `SapOSConfiguration`, `SapSoftwareConfiguration`, `SapSshKeyPair`, `SapSshPublicKey`, `SapStartContent`, `SapStopContent`, and `SapWindowsConfiguration`.

Added support for handling unknown OS and software configurations with `UnknownSapOSConfiguration` and `UnknownSapSoftwareConfiguration`.

Implemented JSON serialization and deserialization for new classes, ensuring proper handling of additional raw data and format validation.

archerzz commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances/CHANGELOG.md:
```diff
@@ -0,0 +1,21 @@
+# Release History
+
+## 1.0.0 (2024-09-01)
```
The release date must be a future date.

jaskisin replied:
@archerzz, this is actually not a releasing date, it is the api version on top of which this sdk is build. Should I put the release date here?

archerzz reviewed (approved):
Need to update the release date in changelog.

ArthurMa1978 commented on sdk/resourcemanager/ci.mgmt.yml:
```diff
@@ -203,6 +203,7 @@ trigger:
     - sdk/websites/Azure.ResourceManager.AppService
     - sdk/workloadmonitor/Azure.ResourceManager.WorkloadMonitor
     - sdk/workloads/Azure.ResourceManager.Workloads
+    - sdk/workloadssapvirtualinstance/Azure.ResourceManager.SapVirtualInstances
```
The approved namespace is `Azure.ResourceManager.WorkloadsSapVirtualInstance`, check here https://github.com/Azure/azure-sdk/issues/7077.

jaskisin replied:
@ArthurMa1978, but the generations(including solution file) are getting created on the basis of the typespec from https://github.com/Azure/azure-rest-api-specs/tree/main/specification/workloads/Workloads.SAPVirtualInstance.Management which is Azure.ResourceManager.SapVirtualInstance.sln.

Is there any way to change this?

ArthurMa1978 reviewed (changes requested):
Need generate API file

ArthurMa1978 committed:
Merge branch 'main' into sap

ArthurMa1978 committed:
update

ArthurMa1978 commented:
/azp run prepare-pipelines

ArthurMa1978 committed:
update

ArthurMa1978 committed:
update

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.ResourceManager.WorkloadsSapVirtualInstance](https://apiview.dev/Assemblies/Review/27832e1d48e2441889de8e1fed359e85?revisionId=27f2691ccc53404da3b7d58de5021466)

ArthurMa1978 committed:
update

ArthurMa1978 committed:
update

ArthurMa1978 commented:
/azp run net - pullrequest

ArthurMa1978 commented:
/azp run net - pullrequest

ArthurMa1978 committed:
update

ArthurMa1978 commented:
/azp run net - pullrequest

ArthurMa1978 commented:
/azp run net - pullrequest

ArthurMa1978 commented:
/azp run net - pullrequest

ArthurMa1978 committed:
Merge branch 'main' into sap

ArthurMa1978 committed:
update

live1206 commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.WorkloadsSapVirtualInstance/src/Generated/SapApplicationServerInstanceData.cs:
```diff
@@ -0,0 +1,156 @@
+        /// <summary> Application server instance SAP Kernel Patch level. </summary>
+        public string KernelPatch { get; }
+        /// <summary> Application server instance SAP IP Address. </summary>
+        public string IPAddress { get; }
```
The type should be IPAddress

live1206 commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.WorkloadsSapVirtualInstance/src/Generated/SapDatabaseInstanceData.cs:
```diff
@@ -0,0 +1,128 @@
+        /// <summary> Database type, that is if the DB is HANA, DB2, Oracle, SAP ASE, Max DB or MS SQL Server. </summary>
+        public string DatabaseType { get; }
+        /// <summary> Database IP Address. </summary>
+        public string IPAddress { get; }
```
The type should be IPAddress

live1206 commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.WorkloadsSapVirtualInstance/src/Generated/Models/EnqueueServerProperties.cs:
```diff
@@ -0,0 +1,77 @@
+        /// <summary> Enqueue Server SAP Hostname. </summary>
+        public string Hostname { get; }
+        /// <summary> Enqueue Server SAP IP Address. </summary>
+        public string IPAddress { get; }
```
The type should be IPAddress

live1206 commented on sdk/workloadssapvirtualinstance/Azure.ResourceManager.WorkloadsSapVirtualInstance/src/Generated/Models/SapVirtualMachineConfiguration.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        /// <summary> The virtual machine size. </summary>
+        public string VmSize { get; set; }
```
should this be `long VmSizeInGB`?

ArthurMa1978 replied:
Not disk size, but scale size like 'Standard_E4ds_v4'

ArthurMa1978 committed:
update

ArthurMa1978 committed:
Merge branch 'main' into sap

ArthurMa1978 commented:
/check-enforcer override
